[ADD] estate:  Created Estate Module

haman-odoo opened the pull request:
This Real Estate module to sell and manage properties from a single place.
It includes property listings, price offers, buyers, agents, tags and property types, all connected together with
features like best-offer calculation, availability, total area and dashboards.

haman-odoo committed:
[ADD] estate: Created base module structure and manifest

Created the initial structure for the Estate module:

- Added __init__.py file to initialize the module.
- Added __manifest__.py file with basic module metadata like:
  - name
  - version
  - depends
  - data
  - description

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/tutorials/pull/1049.png)](https://mergebot.odoo.com/odoo/tutorials/pull/1049)

haman-odoo committed:
[IMP] estate: Added EstateProperty model with all required fields

Chapter 3 : Added EstateProperty model with core fields.
Updated __init__.py files to load new models.

haman-odoo committed:
[IMP] estate:  Added access control for EstateProperty model

Chapter 4 : Added ir.model.access.csv file for defining EstateProperty access rules.
Granted basic read, write, create, and delete permissions to internal users.
Updated __manifest__.py to load the security access file.

haman-odoo committed:
[IMP] estate:  Added actions, menus and first list/form views

Chapter 5 : Added EstateProperty list and form views to display and manage properties.
Created ir.actions.act_window action to open EstateProperty records.
Added Real Estate main menu, submenu, and Properties menu item linked to the action.

haman-odoo committed:
[IMP] estate: Added custom list, form and search views

chapter 6 : Created list, form, and search views for the estate.property model.
Added menu items and actions to open the property views in the Real Estate module.
Created new model estate.property.type and added its list & form views, menu, and access rights.

haman-odoo committed:
[IMP] estate:  Added relations between property, type and offer models

Chapter 7 : Added Many2many - tag_ids field for assigning multiple tags to a property.
Added Many2one - salesperson_id, buyer_id, property_type_id fields.
Added One2many - offer_ids field to show multiple offers linked to one property.

bit-odoo commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,20 @@
+    'category': 'Sales/Real Estate',
+    'installable': True,
+    'auto_install': False,
+    'description': """Description text""",
```
I think we can add description after the name so others can read it easily.

bit-odoo commented on estate/models/__init__.py:
```diff
@@ -0,0 +1,4 @@
+from . import estate_property  # noqa: F401
```
```suggestion
from . import estate_property
```

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,76 @@
+from odoo import models, fields
+from dateutil.relativedelta import relativedelta
+from datetime import date
```
First, we can import external libraries.

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,76 @@
+
+    bedrooms = fields.Integer(string="Bedrooms", default=2)
+    living_area = fields.Integer(string="Living Area (sqm)")
+    facades = fields.Integer(string="Number of Facades")
```
No need to declare a string if it is the same as the field name because Odoo can generate a string.

bit-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,29 @@
+    _description = "Real Estate Property Offer"
+
+    price = fields.Float(string="Price")
+
```
Unneccary diff.

bit-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,29 @@
+        string="Status",
+        copy=False,        # do not copy when duplicating the offer
+    )
+
```
Unneccary diff.

bit-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,29 @@
+        string="Partner",
+        required=True,
+    )
+
```
Unneccary diff.

bit-odoo commented on estate/models/estate_property_tag.py:
```diff
@@ -0,0 +1,8 @@
+    _name = "estate.property.tag"
+    _description = "Real Estate Property Tag"
+
+    name = fields.Char(string="Name", required=True)
```
Same here, no need to write a string.

bit-odoo commented on estate/views/estate_menus.xml:
```diff
@@ -0,0 +1,15 @@
+<!-- /home/odoo/odoo19/tutorials/estate/views/estate_menus.xml -->
```
unneccary diff.

bit-odoo commented on estate/views/estate_property_offer_views.xml:
```diff
@@ -0,0 +1,32 @@
+<!-- /home/odoo/odoo19/tutorials/estate/views/estate_property_offer_views.xml -->
```
unneccary diff.

bit-odoo commented on estate/views/estate_property_type_views.xml:
```diff
@@ -0,0 +1,9 @@
+<!-- /home/odoo/odoo19/tutorials/estate/views/estate_property_type_views.xml -->
```
unneccary diff.

bit-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,124 @@
+<!-- /home/odoo/odoo19/tutorials/estate/views/estate_property_views.xml -->
```
unneccary diff.

bit-odoo reviewed (commented):
Hello @haman-odoo 
You did a good job 👏

I have added some suggestions. And can you please improve your commit message, and pr message, and title regarding these guidelines - https://www.odoo.com/documentation/19.0/contributing/development/git_guidelines.html

One more thing, writing a comment is a good idea. But i think it is good only if our code is complex or doing some magic things, etc.

Thanks

haman-odoo commented:
Hello Sir, I’ve applied all your suggested changes and updated the commit messages.

haman-odoo committed:
[IMP] estate: Compute total area, best price and add onchange for garden fields

Chapter 8 : Added compute, inverse, and onchange methods for dynamic field updates.
Implemented automatic offer price, total area, and deadline calculations.
Added Tag and Offer models with validity and date deadline logic.

haman-odoo committed:
[IMP] estate:  Added action buttons to manage property and offer status

Chapter 9 : Added Sold/Cancel buttons on properties.
Added Accept/Refuse buttons on offers.
Set selling price and buyer when an offer is accepted.

haman-odoo committed:
[IMP] estate:  Added SQL and Python constraints for data validation

Chapter 10 : Added SQL constraints:
  - expected price must be strictly positive
  - selling price must be positive
  - offer price must be strictly positive
  - property type and tag names must be unique
Added a Python constraint to prevent selling prices lower than 90% of the expected price

bit-odoo commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,20 @@
+    'category': 'Sales/Real Estate',
+    'installable': True,
+    'auto_install': False,
+    'description': """Real estate management tutorial module with properties, offers, types and tags.""",
```
Please adapt this -> https://github.com/odoo/tutorials/pull/1049#discussion_r2601666285

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,128 @@
+from datetime import date
+from odoo import models, fields, api
+from odoo.exceptions import UserError,ValidationError
+from odoo.tools.float_utils import float_compare, float_is_zero
```
```suggestion
from dateutil.relativedelta import relativedelta
from datetime import date

from odoo import models, fields, api
from odoo.exceptions import UserError,ValidationError
from odoo.tools.float_utils import float_compare, float_is_zero
```

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,128 @@
+        default=lambda sself: date.today() + relativedelta(months=3),
+        copy=False
+    )
+    expected_price = fields.Float(string="Expected Price", required=True)
```
Please follow this -> https://github.com/odoo/tutorials/pull/1049#discussion_r2601673366

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,128 @@
+                raise UserError("Sold properties cannot be cancelled.")
+            else:
+                rec.state = "cancelled"
+        return True
```
I think `action_cancel_property` will be good.

It's good to make the error message translatable.
`raise exceptions.UserError(_("Sold properties cannot be cancelled."))`

We can use filtered() here.

`if self.filtered(lamda x: x.state=="sold")`

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,128 @@
+                rec.state = "cancelled"
+        return True
+
+    def action_set_sold(self):
+        for rec in self:
+            if rec.state == "cancelled":
+                raise UserError("Canceled properties cannot be sold.")
+            else:
+                rec.state = "sold"
```
Same here.

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,128 @@
+            else:
+                rec.state = "sold"
+        return True
+    @api.constrains("selling_price", "expected_price")
```
Should be above one empty line.

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,128 @@
+            if rec.selling_price == 0:
+                return False
+            if float_compare(rec.selling_price, rec.expected_price * 0.9, precision_digits=2) < 0:
+                raise ValidationError(
```
Same here, it should be translatable.

bit-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,63 @@
+from dateutil.relativedelta import relativedelta
+from odoo import models, fields, api
+from odoo.exceptions import UserError
```
```suggestion
from dateutil.relativedelta import relativedelta

from odoo import models, fields, api
from odoo.exceptions import UserError
```

bit-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,63 @@
+            offer.status = 'refused'
+        return True
+
+    _offer_price = models.Constraint(
```
It is good to declare constraint after the immediate declaration of all fields.

bit-odoo commented on estate/views/estate_menus.xml:
```diff
@@ -0,0 +1,16 @@
+
+    <menuitem id="menu_estate_property_tags" name="Property Tags"
+              parent="menu_estate_settings"
+              action="action_estate_property_tag"/>
```
Indentation issue and empty line before <menuitem>. if you want to leave an empty line, then you should follow the same for others too.

bit-odoo commented on estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,113 @@
+                <field name="selling_price"/>
+                <field name="bedrooms"/>
+                <field name="living_area"/>
+                <field name="date_availability"/>
```
Except `name`,`selling_price`,`expected_price`,`date_availability` we can make other fields optional.

bit-odoo reviewed (commented):
Hello, Good Job!
I have made some suggestions.

Can you please update PR title?

haman-odoo committed:
[IMP] estate:  Added Inline Views, Widgets, List Order, Attributes and Buttons

Chapter 11 : Added inline views, widgets, and stat buttons for clear navigation

Added list order like model, view and manual for easy to find

Added default search filters, editable list views, optional fields

haman-odoo committed:
[IMP] estate:  Added User Inheritance and property listing

Chapter 12 : Add model inheritance to link properties with users
Show available properties on the user form view
Add view inheritance to display properties in a separate notebook tab

haman-odoo committed:
[IMP] estate:  Created estate_account module and generate invoices on sale

Chapter 13 : Create estate_account link module to connect estate and account
Extend property sold action to create a customer invoice
Add invoice lines for commission and administrative fees on sale

haman-odoo committed:
[IMP] estate:  Enhance kanban view with conditional fields and grouping

Chapter 14 : Added expected price, best offer, selling price, and property tags
- Displayed best offer only when an offer is received
- Displayed selling price only when an offer is accepted or sold
- Grouped properties by property type by default
- Disabled drag & drop in kanban view

haman-odoo commented:
Hello Sir, I have applied all you suggestions and PR title.

haman-odoo committed:
[IMP] awesome_owl:  Added Counter and Card component, props and callbacks

Added Counter and extract it into sub component
Added Card component with markup to display html values
Added callback() function, props and incrementSum method in Playground

haman-odoo committed:
[IMP] awesome_owl: Implement TodoList, slots, and interactive Card features

Added TodoList and TodoItem components with reactive state, props validation
Implemented dynamic attributes, checkbox toggling, todo creation, deletion
Introduced DOM refs and lifecycle hooks to autofocus inputs using useRef and onMounted
Refactored Card into a generic slot-based component

haman-odoo committed:
[IMP] awesome_dashboard:  Added layout, buttons and dashboard item

- Integrated Layout component to structure the dashboard with a control panel
- Added Customers and Leads buttons to the control panel using the action
  service for quick navigation
- Implemented a reusable DashboardItem component with slot support

haman-odoo committed:
[IMP] awesome_dashboard:  Add dashboard items, statistics service, and pie chart

- Implemented server-side statistics loading via /awesome_dashboard/statistics
- Implemented a statistics service with memoized RPC calls
- Added a PieChart component with Chart.js to visualize t-shirt sales by size

bit-odoo commented on awesome_dashboard/static/src/dashboard.scss:
```diff
@@ -0,0 +1,3 @@
+.o_dashboard {
+    background-color: #808080;
+}
```
Should be one empty line at the EOF.

bit-odoo commented on awesome_dashboard/static/src/dashboard_item/dashboard_item.js:
```diff
@@ -0,0 +1,16 @@
+    get style(){
+        return `width: ${18 * this.props.size}rem;`;
+    }
+}
```
Should be one empty line at the EOF.

bit-odoo commented on awesome_dashboard/static/src/pie_chart/pie_chart.js:
```diff
@@ -0,0 +1,50 @@
+            },
+        });
+    }
+}
```
Should be one empty line at the EOF.

bit-odoo commented on awesome_owl/static/src/playground.js:
```diff
@@ -1,5 +1,19 @@
+        this.state.sum++;
+    }
 }
+
```
Unneccary diff.

bit-odoo commented on awesome_owl/static/src/todo/todo_item.xml:
```diff
@@ -0,0 +1,17 @@
+        </div>
+    </t>
+
+</templates>
```
Should be one empty line at the EOF.

bit-odoo commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,21 @@
+{
+    'name': "Real Estate",
```
Follow the same quote throughout the file.

bit-odoo commented on estate/__manifest__.py:
```diff
@@ -0,0 +1,21 @@
+    'license': 'LGPL-3',
+    'summary': 'Real Estate advertisement tutorial module',
+    'depends': ['base'],
+    'author': "Harsh Maniya",
```
same.

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,136 @@
+    )
+    best_price = fields.Float(
+        compute="_compute_best_price",
+    )
```
Should be one empty line below it.

bit-odoo commented on estate/models/estate_property.py:
```diff
@@ -0,0 +1,136 @@
+
+    name = fields.Char(required=True)
+    description = fields.Text()
+    postcode = fields.Char()
```
Why char? I think it should be an integer.

bit-odoo commented on estate/models/estate_property_offer.py:
```diff
@@ -0,0 +1,81 @@
+    )
+    property_type_id = fields.Many2one(
+        related="property_id.property_type_id", store=True
+    )
```
Should be one empty line below it.

bit-odoo reviewed (commented):
Hello,
I have made some suggestions. Please do not make same mistake repeatedly.

haman-odoo committed:
[IMP] awesome_dashboard: Added statistics service and lazy components to dashboard

bit-odoo commented on awesome_dashboard/static/src/dashboard/dashboard_items.js:
```diff
@@ -0,0 +1,53 @@
+        data: data.shirt_by_size,
+    }),
+},
+];
```
Should be one empty line at the EOF.

bit-odoo reviewed (commented):
Please do not do same mistake repeatedly. Also please adapt the remaining changes that i have suggested you in previous review.

bit-odoo commented on awesome_dashboard/static/src/dashboard/pie_chart/pie_chart_card.xml:
```diff
@@ -0,0 +1,9 @@
+            <PieChart data="this.props.data"/>
+        </div>
+    </t>
+</templates>
```
Again same mistake.

haman-odoo committed:
[IMP] awesome_dashboard:  Make generic and extensible dashboard architecture

Refactored dashboard to remove hardcoded items.
Implemented dynamic dashboard rendering using components and props.
Enabled extensibility by registering dashboard items in a registry.

haman-odoo committed:
[IMP] awesome_dashboard:  Add extensible dashboard registry and item dialog

Moved dashboard items to a registry for easy extension.
Added settings dialog to show or hide dashboard cards.
Persisted user dashboard preferences using local storage.

haman-odoo commented:
Hello Sir, I have applied all changes that you have suggested and i will do not make same mistake.
